### PR TITLE
[13/N] Enable clang-tidy on headers of torch/csrc

### PR DESCRIPTION
**Pull request by cyyever**
cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/Exceptions.h [hunk 1]
```diff
@@ -385,9 +384,11 @@ auto wrap_pybind_function_impl_(
     HANDLE_TH_ERRORS
     if (release_gil) {
       py::gil_scoped_release no_gil;
-      return c10::guts::invoke(f, std::forward<Arg<Func, Is>>(args)...);
+      return c10::guts::invoke(
+          std::move(f), std::forward<Arg<Func, Is>>(args)...);
```
Comment (Skylion007) [inline torch/csrc/Exceptions.h:388]: Why should this be `std::move` instead of std::forward?

Comment (cyyever) [reply torch/csrc/Exceptions.h:388]: Since f is captured as a value?

File: .lintrunner.toml [hunk 2]
```diff
@@ -244,6 +244,7 @@ include_patterns = [
     'c10/**/*.cpp',
     'c10/core/**/*.h',
     'c10/util/**/*.h',
+    'torch/csrc/*.h',
```
[the remark this answers is not in the document]
Comment (cyyever) [reply .lintrunner.toml]: I want to enable it for torch/csrc/*.h first and then enable it for all headers of torch/csrc. We would consider aten headers finally.

**Review by Skylion007 — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Thanks!

File: torch/csrc/Exceptions.h (hunk 1 above)
Comment (Flamefire) [reply torch/csrc/Exceptions.h:388]: I think the move here is wrong and leads to potential use of a moved-from object:

- `wrap_pybind_function` is used to bind Python functions to C++ functions via pybind.
- The created lambda may be called multiple times
- After the first call `f` in the lambda is technically in a moved-from state and must no longer be used
- But every successive call will use it.
- That C++ function (i.e. `f`) might be anything, e.g. a capturing lambda which after this move might have "lost" a captured resource

I reverted this part of the change in https://github.com/pytorch/pytorch/pull/116695

**Pull request closed without merging**